sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/GraphIdentifierManager.scala

```diff
@@ -0,0 +1,345 @@
+   */
+  @throws[UnsupportedOperationException]
+  def toTableIdentifier(nameParts: Seq[String]): TableIdentifier = {
+    nameParts.length match {
+      case 1 => TableIdentifier(tableName = nameParts.head)
+      case 2 => TableIdentifier(table = nameParts(1), database = Option(nameParts.head))
+      case 3 =>
+        TableIdentifier(
+          table = nameParts(2),
+          database = Option(nameParts(1)),
+          catalog = Option(nameParts.head)
+        )
+      case _ =>
+        throw new UnsupportedOperationException(
+          s"4+ part table identifier ${nameParts.mkString(".")} is not supported."
+        )
+    }
+  }
+
+  /**
+   * Returns the table identifier constructed from the logical plan.
+   *
+   * @param table the logical plan.
+   * @return the table identifier constructed from the logical plan.
+   * @throws SparkException if the table identifier cannot be resolved.
+   */
+  def toTableIdentifier(table: LogicalPlan): TableIdentifier = {
+    val parts = table match {
+      case r: ResolvedIdentifier => r.identifier.namespace.toSeq :+ r.identifier.name
+      case u: UnresolvedIdentifier => u.nameParts
+      case u: UnresolvedRelation => u.multipartIdentifier
+      case _ =>
+        throw new UnsupportedOperationException(s"Unable to resolve name for $table.")
+    }
+    toTableIdentifier(parts)
+  }
+
+  /** Return whether the input identifier is a single-part identifier.  */
+  def isSinglePartIdentifier(identifier: TableIdentifier): Boolean = {
+    identifier.database.isEmpty && identifier.catalog.isEmpty
+  }
+
+  /**
+   * Return true if the identifier should be resolved as a path-based reference
+   * (i.e., `datasource`.`path`).
+   */
+  def isPathIdentifier(spark: SparkSession, identifier: TableIdentifier): Boolean = {
+    if (identifier.nameParts.length != 2) {
+      return false
+    }
+    val Seq(datasource, path) = identifier.nameParts
+    val sqlConf = spark.sessionState.conf
+
+    def isDatasourceValid = {
+      try {
+        DataSource.lookupDataSource(datasource, sqlConf)
+        true
+      } catch {
+        case _: ClassNotFoundException => false
+      }
+    }
+
+    // Whether the provided datasource is valid.
+    isDatasourceValid
+  }
+
+  /** Fully qualifies provided identifier with provided catalog & schema. */
+  def fullyQualifyIdentifier(
+      maybeFullyQualifiedIdentifier: TableIdentifier,
+      currentCatalog: Option[String],
+      currentDatabase: Option[String]
+  ): TableIdentifier = {
+    maybeFullyQualifiedIdentifier.copy(
+      database = maybeFullyQualifiedIdentifier.database.orElse(currentDatabase),
+      catalog = maybeFullyQualifiedIdentifier.catalog.orElse(currentCatalog)
+    )
+  }
+
+  /** Assert whether the identifier is properly fully qualified when creating a dataset. */
+  def assertIsFullyQualifiedForCreate(identifier: TableIdentifier): Unit = {
+    // TODO: validate catalog exists
```

Comment:
Nitpick: Can we take this TODO out? Even if validating that the catalog exists would be nice, better to track this with a JIRA than to use the codebase as an issue tracker.